pkg/services/chartmuseum_client.go

```diff
@@ -0,0 +1,177 @@
+				return nil, err
+			}
+
+			chartInstanceVersion := chartInstanceContainerMap["version"].Data().(string)
+			if _, chartInstanceExists := c.ChartVersionCache[chartName]; !chartInstanceExists {
+				c.ChartVersionCache[chartName] = make(map[string]bool)
+			}
+			c.ChartVersionCache[chartName][chartInstanceVersion] = true
+		}
+	}
+
+	return c, nil
+}
+
+func (c *ChartmuseumClient) IsActive() bool {
+	return c.ChartmuseumUrl != ""
+}
+
+func (c *ChartmuseumClient) HasBasicAuth() bool {
+	return c.ChartmuseumUsername != "" && c.ChartmuseumPassword != ""
+}
+
+func (c *ChartmuseumClient) GetAllCharts() ([]byte, error) {
+	req, err := http.NewRequest("GET", fmt.Sprintf("%s/api/charts", c.ChartmuseumUrl), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.HasBasicAuth() {
+		req.SetBasicAuth(c.ChartmuseumUsername, c.ChartmuseumPassword)
+	}
+
+	resp, err := c.HttpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != 200 {
+		return nil, errors.New(fmt.Sprintf("Receiving list of charts failed. Status code - %d, Body - %s", resp.StatusCode, string(respBody)))
+	}
+
+	return respBody, nil
+}
+
+func (c *ChartmuseumClient) Upload(chartName string, chartVersion string, f *os.File) error {
+	fileContents, err := ioutil.ReadAll(f)
+	if err != nil {
+		return err
+	}
+	fi, err := f.Stat()
+	if err != nil {
+		return err
+	}
+	f.Close()
+	if err != nil {
+		return err
+	}
+
+	body := new(bytes.Buffer)
+	writer := multipart.NewWriter(body)
+	part, err := writer.CreateFormFile("chart", fi.Name())
+	if err != nil {
+		return err
+	}
+	_, err = part.Write(fileContents)
+	if err != nil {
+		return err
+	}
+	err = writer.Close()
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest("POST", fmt.Sprintf("%s/api/charts", c.ChartmuseumUrl), body)
+	if err != nil {
+		return err
+	}
+
+	if c.HasBasicAuth() {
+		req.SetBasicAuth(c.ChartmuseumUsername, c.ChartmuseumPassword)
+	}
+	req.Header.Add("Content-Type", writer.FormDataContentType())
+
+	resp, err := c.HttpClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	responseBody, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != 201 {
+		return errors.New(fmt.Sprintf("Receiving list of charts failed. Status code - %d, Body - %s", resp.StatusCode, string(responseBody)))
+	}
+
+	if _, chartExists := c.ChartVersionCache[chartName]; !chartExists {
+		c.ChartVersionCache[chartName] = make(map[string]bool)
+	}
+
+	c.ChartVersionCache[chartName][chartVersion] = true
+
+	zap.L().Sugar().Infof("Successfully uploaded chart: %s-%s", chartName, chartVersion)
+
+	return resp.Body.Close()
+}
+
+func (c *ChartmuseumClient) IsExists(chartName string, chartVersion string) bool {
+	if _, chartExists := c.ChartVersionCache[chartName]; !chartExists {
+		return false
+	}
+	return c.ChartVersionCache[chartName][chartVersion]
```

Comment:
Didn't understand. I'm using this method in line 173 to check if chart in cache exists, then since each chart contains `map[string]bool`, we can easily check it by specifying version. I checked in go playground and it's working:
```
package main

import "fmt"

func main() {
	cachemap := make(map[string]bool)
	cachemap["1.0.3"] = true
	fmt.Println(cachemap["1.0.2"]) // false
}
```

Correct me if I'm wrong